mapiproxy/servers/default/asyncemsmdb/dcesrv_asyncemsmdb.c

```diff
@@ -0,0 +1,1230 @@
+
+	ndr_err_code = ndr_pull_mapistore_notification(ndr_pull, NDR_SCALARS, &n);
+	nn_freemsg(str);
+	if (ndr_err_code != NDR_ERR_SUCCESS) {
+		OC_DEBUG(0, "[asyncemsmdb]: Invalid mapistore_notification structure");
+		talloc_free(mem_ctx);
+		return;
+	}
+
+	/* Sanity checks on notification payload */
+	if (n.vnum >= MAPISTORE_NOTIFICATION_VMAX) {
+		OC_DEBUG(0, "[asyncemsmdb]: Invalid version, expected at max %d but got %d",
+			 MAPISTORE_NOTIFICATION_VMAX - 1, n.vnum);
+		talloc_free(mem_ctx);
+		return;
+	}
+
+	ndr_print = talloc_zero(mem_ctx, struct ndr_print);
+	if (!ndr_print) {
+		OC_DEBUG(0, "[asyncemsmdb]: No more memory");
+		talloc_free(mem_ctx);
+		return;
+	}
+	ndr_print->depth = 1;
+	ndr_print->print = ndr_print_debug_helper;
+	ndr_print->no_newline = false;
+
+	ndr_print_mapistore_notification(ndr_print, "notification", &n);
+
+	OC_DEBUG(5, "Notification received for session: %s", p->emsmdb_session_str);
+
+	retval = mapistore_notification_subscription_get(mem_ctx, p->mstore_ctx, p->emsmdb_uuid, &r);
+	if (retval != MAPISTORE_SUCCESS) {
+		OC_DEBUG(0, "no subscription to process");
+		talloc_free(mem_ctx);
+		return;
+	}
+
+	OC_DEBUG(5, "%d subscriptions available:", r.v.v1.count);
+	ndr_print_mapistore_notification_subscription(ndr_print, "subscriptions", &r);
+	talloc_free(ndr_print);
+
+	/* Process notifications */
+	switch (n.v.v1.flags) {
+	case (sub_NewMail):
+		ret = process_newmail_notification(mem_ctx, p, &n, &r);
+		if (ret) {
+			OC_DEBUG(0, "[asyncemsmdb]: Failed to process newmail notification (error=0x%x)", ret);
+			talloc_free(mem_ctx);
+			return;
+		}
+		break;
+	default:
+		OC_DEBUG(0, "[asyncemsmdb]: Unsupported notification 0x%x", n.v.v1.flags);
+		talloc_free(mem_ctx);
+		return;
+	}
+	talloc_free(ndr_pull);
+	talloc_free(mem_ctx);
+
+	p->r->out.pulFlagsOut = talloc_zero(p->dce_call, uint32_t);
+	*p->r->out.pulFlagsOut = 0x1;
+
+	status = dcesrv_reply(p->dce_call);
+	if (!NT_STATUS_IS_OK(status)) {
+		OC_DEBUG(0, "EcDoAsyncWaitEx_handler: dcesrv_reply() failed - %s", nt_errstr(status));
+	}
+
+	talloc_free(p->fd_event);
+	p->fd_event = NULL;
+}
+
+
+ /**
+    \details exchange_async_emsmdb EcDoAsyncWaitEx (0x0) function
+
+   \param dce_call pointer to the session context
+   \param mem_ctx pointer to the memory context
+   \param r pointer to the EcDoAsyncWaitEx request data
+
+   \return NT_STATUS_OK on success
+ */
+static NTSTATUS dcesrv_EcDoAsyncWaitEx(struct dcesrv_call_state *dce_call,
+				       TALLOC_CTX *mem_ctx,
+				       struct EcDoAsyncWaitEx *r)
+{
+	enum mapistore_error			retval;
+	struct exchange_asyncemsmdb_session	*session = NULL;
+	struct asyncemsmdb_private_data		*p = NULL;
+	struct mapistore_context		*mstore_ctx = NULL;
+	struct GUID				uuid;
+	char					*cn = NULL;
+	const char				*ip_addr = NULL;
+	int					nn_retval;
+	size_t					sz = 0;
+	char					*bind_addr = NULL;
+	int					port = 0;
+
+	OC_DEBUG(0, "exchange_asyncemsmdb: EcDoAsyncWaitEx (0x0)");
```

Comment:
Sorry for being picky, but we should use the same level (3) that we use in `dcesrv_exchange_emsmdb.c`.